packages/core/src/execution-engine/workflow-execute.ts

```diff
@@ -1089,72 +1089,68 @@ export class WorkflowExecute {
+		if (inputData.hasOwnProperty('main') && inputData.main.length > 0) {
+			// If the node is disabled simply return the data from the first main input
+			if (inputData.main[0] === null) {
+				return { data: undefined };
 			}
-			return { data: undefined };
+			return { data: [inputData.main[0]] };
 		}
+		return { data: undefined };
+	}
 
-		const nodeType = workflow.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
-
-		const isDeclarativeNode = nodeType.description.requestDefaults !== undefined;
-
-		const customOperation = this.getCustomOperation(node, nodeType);
-
-		let connectionInputData: INodeExecutionData[] = [];
+	/**
+	 * Prepares input data for node execution based on node type and workflow settings
+	 */
+	private prepareConnectionInputData(
+		workflow: Workflow,
+		nodeType: INodeType,
+		customOperation: ReturnType<WorkflowExecute['getCustomOperation']>,
+		inputData: ITaskDataConnections,
+	): INodeExecutionData[] | null {
 		if (
-			nodeType.execute ||
-			customOperation ||
-			(!nodeType.poll && !nodeType.trigger && !nodeType.webhook)
+			!nodeType.execute &&
+			!customOperation &&
+			(nodeType.poll || nodeType.trigger || nodeType.webhook)
 		) {
-			// Only stop if first input is empty for execute runs. For all others run anyways
-			// because then it is a trigger node. As they only pass data through and so the input-data
-			// becomes output-data it has to be possible.
+			// For poll, trigger, and webhook nodes, we don't need to process input data
+			return [];
+		}
 
-			if (inputData.main?.length > 0) {
-				// We always use the data of main input and the first input for execute
-				connectionInputData = inputData.main[0] as INodeExecutionData[];
-			}
+		if (!inputData.main?.length) {
+			return null;
+		}
 
-			const forceInputNodeExecution = workflow.settings.executionOrder !== 'v1';
-			if (!forceInputNodeExecution) {
-				// If the nodes do not get force executed data of some inputs may be missing
-				// for that reason do we use the data of the first one that contains any
-				for (const mainData of inputData.main) {
-					if (mainData?.length) {
-						connectionInputData = mainData;
-						break;
-					}
+		// We always use the data of main input and the first input for execute
+		let connectionInputData = inputData.main[0] as INodeExecutionData[];
+
+		const forceInputNodeExecution = workflow.settings.executionOrder !== 'v1';
+		if (!forceInputNodeExecution) {
+			// If the nodes do not get force executed data of some inputs may be missing
+			// for that reason do we use the data of the first one that contains any
+			for (const mainData of inputData.main) {
+				if (mainData?.length) {
+					connectionInputData = mainData;
+					break;
 				}
 			}
+		}
 
-			if (connectionInputData.length === 0) {
-				// No data for node so return
-				return { data: undefined };
-			}
+		if (connectionInputData.length === 0) {
+			return null;
```

Comment:
Missing context so this might be intended, but I notice the return value differs?